ramalama/gguf_parser.py

```diff
@@ -0,0 +1,201 @@
+    GGUFValueType.UINT8: "B",
+    GGUFValueType.INT8: "b",
+    GGUFValueType.UINT16: "H",
+    GGUFValueType.INT16: "h",
+    GGUFValueType.UINT32: "I",
+    GGUFValueType.INT32: "i",
+    GGUFValueType.FLOAT32: "f",
+    GGUFValueType.BOOL: "?",
+    GGUFValueType.UINT64: "Q",
+    GGUFValueType.INT64: "q",
+    GGUFValueType.FLOAT64: "d",
+}
+
+GGUF_NUMBER_FORMATS: list[GGUFValueType] = [
+    GGUFValueType.UINT8,
+    GGUFValueType.INT8,
+    GGUFValueType.UINT16,
+    GGUFValueType.INT16,
+    GGUFValueType.UINT32,
+    GGUFValueType.INT32,
+    GGUFValueType.FLOAT32,
+    GGUFValueType.UINT64,
+    GGUFValueType.INT64,
+    GGUFValueType.FLOAT64,
+]
+
+
+class ParseError(Exception):
+    pass
+
+
+class GGUFInfoParser:
+
+    def is_model_gguf(model: Model, cli_args) -> bool:
+        try:
+            with open(model.get_model_path(cli_args), "rb") as model_file:
+                magic_number = GGUFInfoParser.read_string(model_file, 4)
+                return magic_number == GGUFModelInfo.MAGIC_NUMBER
+        except Exception as ex:
+            console.warning(f" Failed to read model '{model}': {ex}")
+            return False
+
+    @staticmethod
+    def read_string(model: io.BufferedReader, length: int = -1) -> str:
+        if length == -1:
+            type_string = GGUF_VALUE_TYPE_FORMAT[GGUFValueType.UINT64]
+            length = struct.unpack(type_string, model.read(struct.calcsize(type_string)))[0]
+        return model.read(length).decode("utf-8")
+
+    @staticmethod
+    def read_number(model: io.BufferedReader, value_type: GGUFValueType, model_uses_little_endian: bool) -> float:
+        if value_type not in GGUF_NUMBER_FORMATS:
+            raise ParseError(f"Value type '{value_type}' not in format dict")
+        typestring = f"{'<' if model_uses_little_endian else '>'}{GGUF_VALUE_TYPE_FORMAT[value_type]}"
+        return struct.unpack(typestring, model.read(struct.calcsize(typestring)))[0]
+
+    @staticmethod
+    def read_bool(model: io.BufferedReader, model_uses_little_endian: bool) -> bool:
+        typestring = f"{'<' if model_uses_little_endian else '>'}{GGUF_VALUE_TYPE_FORMAT[GGUFValueType.BOOL]}"
+        value = struct.unpack(typestring, model.read(struct.calcsize(typestring)))[0]
+        if value not in [0, 1]:
+            raise ParseError(f"Invalid bool value '{value}'")
+        return value == 1
+
+    @staticmethod
+    def read_value_type(model: io.BufferedReader, model_uses_little_endian: bool) -> GGUFValueType:
+        value_type = GGUFInfoParser.read_number(model, GGUFValueType.UINT32, model_uses_little_endian)
+        return GGUFValueType(value_type)
+
+    @staticmethod
+    def read_value(model: io.BufferedReader, value_type: GGUFValueType, model_uses_little_endian: bool) -> Any:
+        if value_type in GGUF_NUMBER_FORMATS:
+            return GGUFInfoParser.read_number(model, value_type, model_uses_little_endian)
+        elif value_type == GGUFValueType.BOOL:
```

Comment:
Not a fan of elif, when the previous if returned a value.  I think this makes the code look confusing.

Reply:
Change it,